storage/innobase/buf/buf0lru.cc

```diff
@@ -385,144 +385,71 @@ we put it to free list to be used.
-		}
-		const bool wake = buf_pool.need_LRU_eviction();
-		mysql_mutex_unlock(&buf_pool.mutex);
-		mysql_mutex_lock(&buf_pool.flush_list_mutex);
-		const auto n_flush = buf_pool.n_flush();
-		if (wake && !buf_pool.page_cleaner_active()) {
-			buf_pool.page_cleaner_wakeup(true);
-		}
-		mysql_mutex_unlock(&buf_pool.flush_list_mutex);
-		mysql_mutex_lock(&buf_pool.mutex);
-		if (!n_flush) {
-			goto not_found;
-		}
-		if (!buf_pool.try_LRU_scan) {
-			my_cond_wait(&buf_pool.done_free,
-				     &buf_pool.mutex.m_mutex);
-		}
-	}
+    block->page.zip.clear();
+    return block;
+  }
 
-not_found:
-	if (n_iterations > 1) {
-		MONITOR_INC( MONITOR_LRU_GET_FREE_WAITS );
-	}
+  MONITOR_INC(MONITOR_LRU_GET_FREE_LOOPS);
+  if (waited || buf_pool.try_LRU_scan)
+  {
+    /* If no block was in the free list, search from the end of the
+    LRU list and try to free a block there.  If we are doing for the
+    first time we'll scan only tail of the LRU list otherwise we scan
+    the whole LRU list. */
+    if (buf_LRU_scan_and_free_block(waited ? ULINT_UNDEFINED : 100))
+      goto retry;
+
+    /* Tell other threads that there is no point in scanning the LRU
+    list. */
+    buf_pool.try_LRU_scan= false;
+  }
 
-	if (n_iterations == 21
-	    && srv_buf_pool_old_size == srv_buf_pool_size
-	    && buf_pool.LRU_warned.test_and_set(std::memory_order_acquire)) {
-		IF_DBUG(buf_lru_free_blocks_error_printed = true,);
-		mysql_mutex_unlock(&buf_pool.mutex);
-		ib::warn() << "Difficult to find free blocks in the buffer pool"
-			" (" << n_iterations << " search iterations)! "
-			<< flush_failures << " failed attempts to"
-			" flush a page!"
-			" Consider increasing innodb_buffer_pool_size."
-			" Pending flushes (fsync) log: "
-			<< log_sys.get_pending_flushes()
-			<< "; buffer pool: "
-			<< fil_n_pending_tablespace_flushes
-			<< ". " << os_n_file_reads << " OS file reads, "
-			<< os_n_file_writes << " OS file writes, "
-			<< os_n_fsyncs
-			<< " OS fsyncs.";
-		mysql_mutex_lock(&buf_pool.mutex);
-	}
+  waited= true;
 
-	/* No free block was found: try to flush the LRU list.
-	The freed blocks will be up for grabs for all threads.
+  while (!(block= buf_LRU_get_free_only()))
+  {
+    buf_pool.stat.LRU_waits++;
 
-	TODO: A more elegant way would have been to return one freed
-	up block to the caller here but the code that deals with
-	removing the block from buf_pool.page_hash and buf_pool.LRU is fairly
-	involved (particularly in case of ROW_FORMAT=COMPRESSED pages). We
-	can do that in a separate patch sometime in future. */
-#ifndef DBUG_OFF
-flush_lru:
-#endif
-	if (!buf_flush_LRU(innodb_lru_flush_size, true)) {
-		MONITOR_INC(MONITOR_LRU_SINGLE_FLUSH_FAILURE_COUNT);
-		++flush_failures;
-	}
+    mysql_mutex_lock(&buf_pool.flush_list_mutex);
+    if (!buf_pool.page_cleaner_active())
+      buf_pool.page_cleaner_wakeup(true);
+    mysql_mutex_unlock(&buf_pool.flush_list_mutex);
+    my_cond_wait(&buf_pool.done_free, &buf_pool.mutex.m_mutex);
```

Comment:
I agree that this could be useful.

Reply:
d45b7f665513e5a8e24799157b64caa7c5685a03 makes the interval 5 seconds, because the `buf_flush_page_cleaner()` is normally scheduled to run every 1 second. We will also attempt to wake up `buf_flush_page_cleaner()` every 5 seconds. The message output was refactored to a new member function `buf_pool.LRU_warn()`.

Reply:
Right, running out of buffer pool is an existing design issue and there is no easy fix for it. With a small buffer pool, it is straightforward to generate a workload where several threads are trying to allocate a page from the buffer pool while the entire buffer pool is buffer-fixed by the active threads. We can try to prevent hangs in some special cases, such as in `trx_purge_attach_undo_recs()` where we limit the purge batch size.

A page allocation failure could be treated in a similar way as an attempt to read a corrupted page: roll back the current operation. But, rollback would need to access some pages as well. Maybe some allocation requests (such as those during rollback) should be treated at a higher priority?